src/core/ddsi/src/ddsi_security_omg.c

```diff
@@ -258,97 +296,162 @@ void q_omg_log_exception(const struct ddsrt_log_cfg *lc, uint32_t cat, DDS_Secur
 
-  match = ddsrt_malloc(sizeof(*match));
-  match->guids.src = *src;
-  match->guids.dst = *dst;
-  match->matched = false;
-  match->crypto_handle = 0;
-  match->tokens = NULL;
+  ddsrt_mutex_lock(&index->lock);
+  if ((match = ddsrt_avl_lookup_ipath(&pending_match_index_treedef, &index->pending_matches, &guids, &ipath)) == NULL)
+  {
+    match = ddsrt_malloc(sizeof(*match));
+    match->crypto_handle = 0;
+    match->tokens = NULL;
+    match->guids = guids;
+    match->kind = kind;
+    ddsrt_avl_insert_ipath(&pending_match_index_treedef, &index->pending_matches, match, &ipath);
+  }
+
+  if (crypto_handle)
+    match->crypto_handle = crypto_handle;
+
+  if (tokens)
+  {
+    match->tokens = tokens;
+    match->expiry = ddsrt_mtime_add_duration(ddsrt_time_monotonic(), DDS_SECS(PENDING_MATCH_EXPIRY_TIME));
+    if (index->evt == NULL)
+      index->evt = qxev_callback(index->gv->xevents, match->expiry, pending_match_expiry_cb, index);
+    else
+      (void)resched_xevent_if_earlier(index->evt, match->expiry);
+  }
+  ddsrt_mutex_unlock(&index->lock);
 
   return match;
 }
 
-static void entity_match_free(struct security_entity_match *match)
+static void clear_and_free_pending_match(dds_security_context *sc, struct pending_match *match)
 {
-  if (match) {
-    if (match->tokens)
-      DDS_Security_ParticipantCryptoTokenSeq_free(match->tokens);
-    ddsrt_free(match);
+  if (match->crypto_handle != 0)
+  {
+    DDS_Security_SecurityException exception = DDS_SECURITY_EXCEPTION_INIT;
+    const char *ename;
+    bool r = true;
+
+    switch (match->kind)
+    {
+    case EK_PROXY_PARTICIPANT:
+      break;
+    case EK_PROXY_READER:
+      ename = "reader";
+      r = sc->crypto_context->crypto_key_factory->unregister_datareader(sc->crypto_context->crypto_key_factory, match->crypto_handle, &exception);
+      break;
+    case EK_PROXY_WRITER:
+      ename = "writer";
+      r = sc->crypto_context->crypto_key_factory->unregister_datawriter(sc->crypto_context->crypto_key_factory, match->crypto_handle, &exception);
+      break;
+    default:
+      assert(0);
+      break;
+    }
+    if (!r)
+      EXCEPTION_ERROR(sc, &exception, "Failed to unregister remote %s crypto "PGUIDFMT" related to "PGUIDFMT, ename, PGUID(match->guids.remote_guid), PGUID(match->guids.local_guid));
   }
+  free_pending_match(match);
 }
 
-static struct security_entity_match * find_entity_match_locked(struct dds_security_match_index *list, const ddsi_guid_t *src, const ddsi_guid_t *dst)
+static void remove_pending_match(struct pending_match_index *index, struct pending_match *match)
 {
-  struct guid_pair guids;
-
-  guids.src = *src;
-  guids.dst = *dst;
+  struct pending_match *found;
+  ddsrt_avl_dpath_t dpath;
 
-  return ddsrt_avl_lookup(&entity_match_treedef, &list->matches, &guids);
+  ddsrt_mutex_lock(&index->lock);
+  if ((found = ddsrt_avl_lookup_dpath(&pending_match_index_treedef, &index->pending_matches, &match->guids, &dpath)) != NULL)
```

Comment:
This lookup bothers me. If the locking and allocation/freeing strategies are sound it appears there should be no possibility of it not being in the index. I can see ``pending_match_expiry_cb`` possibly occurring just before locking ``index->lock``, but then that really means the caller's activities immediately preceding this function are a potential use-after-free.